src/libstd/ffi/c_str.rs

```diff
@@ -718,6 +719,77 @@ impl From<Box<CStr>> for CString {
     }
 }
 
+#[stable(feature = "cstring_from_vec_of_nonzerou8", since = "1.39.0")]
+impl From<Vec<NonZeroU8>> for CString {
+    /// Converts a [`Vec`]`<`[`NonZeroU8`]`>` into a [`CString`] without
+    /// copying nor checking for inner null bytes.
+    ///
+    /// [`CString`]: ../ffi/struct.CString.html
+    /// [`NonZeroU8`]: ../num/struct.NonZeroU8.html
+    /// [`Vec`]: ../vec/struct.Vec.html
+    #[inline]
+    fn from(v: Vec<NonZeroU8>) -> CString {
+        /// # Safety
+        ///
+        ///   - for any `length: usize`, it must be sound to transmute between
+        ///     `[T; length]` and `[U; length]`.
+        ///
+        ///   - for any `cap: usize`, `Layout<[T; cap]>` needs to be equal to
+        ///     `Layout<[U; cap]>` (for the allocator).
+        #[inline]
+        unsafe fn transmute_vec<T, U>(v: Vec<T>) -> Vec<U> {
+            // Necessary condition for `Layout<[T; N]> == Layout<[U; N]>`:
+            assert_eq!(
+                crate::alloc::Layout::new::<T>(),
+                crate::alloc::Layout::new::<U>(),
+            );
+            // The previous assert should imply the following one:
+            debug_assert_eq!(
+                crate::alloc::Layout::array::<T>(v.capacity()),
+                crate::alloc::Layout::array::<U>(v.capacity()),
+            );
+            // # Safety
+            //
+            //   - this is the pattern blessed by transmute's documentation
+            //     (and replacing `mem::forget` by `ManuallyDrop`, _c.f._,
+            //     issue #62553).
+            //
+            //   - `transmute_vec` contract guarantees sound transmuting
+            //     between `[T; length]` and `[U; length]`.
+            let length: usize = v.len();
+            let capacity: usize = v.capacity();
+            let ptr: *mut T = {
+                // End-of-scope `mem::forget` but without aliasing problems.
+                let mut v = mem::ManuallyDrop::<Vec<T>>::new(v);
+                let v: &mut Vec<T> = &mut *v;
+                v.as_mut_ptr() // Cannot be aliased.
```

Comment:
`ManuallyDrop` implements `Deref`; why is the intermediate `let v: &mut Vec<T> = ...` needed?

Reply:
It's just me being (overly?) cautious with type inference / method resolution within `unsafe` code , so I wanted to have that `Deref` be very explicit.